File: backend/src/repositories/in-memory-project.repository.ts
import { LoopProject } from '../models/project';
import { ProjectRepository } from './project.repository';

export class InMemoryProjectRepository implements ProjectRepository {
  private projects = new Map<string, LoopProject>();

  async listProjects() {
    return Array.from(this.projects.values()).sort((a, b) =>
      b.updatedAt.localeCompare(a.updatedAt)
    );
  }

  async getProjectById(projectId: string) {
    return this.projects.get(projectId) ?? null;
  }

  async createProject(project: LoopProject) {
    this.projects.set(project.id, project);

    return project;
  }

  async reset() {
    this.projects.clear();
  }
}

export const projectRepository = new InMemoryProjectRepository();

File: backend-dotnet/src/Loopr.Api/Repositories/InMemoryProjectRepository.cs
using Loopr.Api.Domain;

namespace Loopr.Api.Repositories;

public sealed class InMemoryProjectRepository : IProjectRepository
{
    private readonly Dictionary<string, LoopProject> projects = new(StringComparer.Ordinal);
    private readonly object syncRoot = new();

    public Task<IReadOnlyList<LoopProject>> ListAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            return Task.FromResult<IReadOnlyList<LoopProject>>(
                projects.Values.OrderByDescending(project => project.UpdatedAt).ToArray()
            );
        }
    }

    public Task<LoopProject?> GetByIdAsync(string projectId, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            projects.TryGetValue(projectId, out var project);

            return Task.FromResult(project);
        }
    }

    public Task<LoopProject> CreateAsync(
        LoopProject project,
        CancellationToken cancellationToken
    )
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            projects[project.Id] = project;
        }

        return Task.FromResult(project);
    }

    public Task ResetAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            projects.Clear();
        }

        return Task.CompletedTask;
    }
}
